File: src/moltblock/graph_schema.py
"""Agent graph schema: DAG of nodes (role + model_binding) and edges (data flow)."""

from pathlib import Path

from pydantic import BaseModel, ConfigDict, Field


class GraphNode(BaseModel):
    """One node in the agent graph: id, role, and which model binding to use."""

    id: str = Field(description="Unique node id (e.g. generator, critic, judge)")
    role: str = Field(description="Role name: generator, critic, judge, router, etc.")
    binding: str = Field(description="Key into bindings dict (e.g. generator, critic)")


class GraphEdge(BaseModel):
    """Edge: from node output -> to node input."""

    model_config = ConfigDict(populate_by_name=True)

    from_node: str = Field(alias="from", description="Source node id")
    to_node: str = Field(alias="to", description="Target node id")


class AgentGraph(BaseModel):
    """Declarative agent graph: nodes and edges. Verifier runs on final node(s) output."""

    nodes: list[GraphNode] = Field(default_factory=list)
    edges: list[GraphEdge] = Field(default_factory=list)
    final_node: str | None = Field(
        default=None,
        description="Node whose output is the final candidate for verification (default: node with no outgoing edges)",
    )

    def predecessors(self, node_id: str) -> list[str]:
        """Return list of node ids that have an edge into node_id."""
        return [e.from_node for e in self.edges if e.to_node == node_id]

    def successors(self, node_id: str) -> list[str]:
        """Return list of node ids that node_id has an edge to."""
        return [e.to_node for e in self.edges if e.from_node == node_id]

    def topological_order(self) -> list[str]:
        """Return node ids in topological order (inputs before outputs)."""
        node_ids = {n.id for n in self.nodes}
        in_degree = {nid: 0 for nid in node_ids}
        for e in self.edges:
            if e.to_node in node_ids:
                in_degree[e.to_node] += 1
        order: list[str] = []
        while len(order) < len(node_ids):
            ready = [nid for nid in node_ids if nid not in order and in_degree[nid] == 0]
            if not ready:
                break
            order.extend(ready)
            for nid in ready:
                for s in self.successors(nid):
                    in_degree[s] -= 1
        return order

    def get_final_node_id(self) -> str | None:
        """Node whose output goes to verifier. Explicit final_node or single node with no outgoing edges."""
        if self.final_node:
            return self.final_node
        candidates = [n.id for n in self.nodes if not self.successors(n.id)]
        return candidates[0] if len(candidates) == 1 else None

    @classmethod
    def load(cls, path: str | Path) -> "AgentGraph":
        """Load graph from JSON or YAML file."""
        path = Path(path)
        if not path.exists():
            raise FileNotFoundError(path)
        raw = path.read_text(encoding="utf-8")
        if path.suffix.lower() in (".yaml", ".yml"):
            try:
                import yaml
                data = yaml.safe_load(raw)
            except ImportError:
                raise ImportError("PyYAML required for YAML graphs: pip install pyyaml")
        else:
            import json
            data = json.loads(raw)
        return cls.model_validate(data)

File: src/graph-schema.ts
/**
 * Agent graph schema: DAG of nodes (role + model_binding) and edges (data flow).
 */

import fs from "node:fs";
import path from "node:path";
import { z } from "zod";

export const GraphNodeSchema = z.object({
  id: z.string().describe("Unique node id (e.g. generator, critic, judge)"),
  role: z.string().describe("Role name: generator, critic, judge, router, etc."),
  binding: z.string().describe("Key into bindings dict (e.g. generator, critic)"),
});

export type GraphNode = z.infer<typeof GraphNodeSchema>;

export const GraphEdgeSchema = z.object({
  from: z.string().describe("Source node id"),
  to: z.string().describe("Target node id"),
});

export type GraphEdge = z.infer<typeof GraphEdgeSchema>;

export const AgentGraphSchema = z.object({
  nodes: z.array(GraphNodeSchema).default([]),
  edges: z.array(GraphEdgeSchema).default([]),
  final_node: z
    .string()
    .nullable()
    .optional()
    .describe(
      "Node whose output is the final candidate for verification (default: node with no outgoing edges)"
    ),
});

export type AgentGraphData = z.infer<typeof AgentGraphSchema>;

/**
 * Declarative agent graph: nodes and edges. Verifier runs on final node(s) output.
 */
export class AgentGraph {
  readonly nodes: GraphNode[];
  readonly edges: GraphEdge[];
  readonly finalNode: string | null;

  constructor(data: AgentGraphData) {
    this.nodes = data.nodes;
    this.edges = data.edges;
    this.finalNode = data.final_node ?? null;
  }

  /**
   * Return list of node ids that have an edge into nodeId.
   */
  predecessors(nodeId: string): string[] {
    return this.edges.filter((e) => e.to === nodeId).map((e) => e.from);
  }

  /**
   * Return list of node ids that nodeId has an edge to.
   */
  successors(nodeId: string): string[] {
    return this.edges.filter((e) => e.from === nodeId).map((e) => e.to);
  }

  /**
   * Return node ids in topological order (inputs before outputs).
   */
  topologicalOrder(): string[] {
    const nodeIds = new Set(this.nodes.map((n) => n.id));
    const inDegree: Record<string, number> = {};
    for (const nid of nodeIds) {
      inDegree[nid] = 0;
    }
    for (const e of this.edges) {
      if (nodeIds.has(e.to)) {
        inDegree[e.to] = (inDegree[e.to] ?? 0) + 1;
      }
    }

    const order: string[] = [];
    while (order.length < nodeIds.size) {
      const ready = [...nodeIds].filter(
        (nid) => !order.includes(nid) && inDegree[nid] === 0
      );
      if (ready.length === 0) {
        break; // cycle or done
      }
      order.push(...ready);
      for (const nid of ready) {
        for (const s of this.successors(nid)) {
          if (inDegree[s] !== undefined) {
            inDegree[s]--;
          }
        }
      }
    }
    return order;
  }

  /**
   * Node whose output goes to verifier. Explicit final_node or single node with no outgoing edges.
   */
  getFinalNodeId(): string | null {
    if (this.finalNode) {
      return this.finalNode;
    }
    const candidates = this.nodes
      .filter((n) => this.successors(n.id).length === 0)
      .map((n) => n.id);
    return candidates.length === 1 ? (candidates[0] ?? null) : null;
  }

  /**
   * Serialize to JSON string.
   */
  toJSON(): string {
    return JSON.stringify({
      nodes: this.nodes,
      edges: this.edges,
      final_node: this.finalNode,
    });
  }

  /**
   * Load graph from JSON or YAML file.
   */
  static load(filePath: string): AgentGraph {
    const resolved = path.resolve(filePath);
    if (!fs.existsSync(resolved)) {
      throw new Error(`Graph file not found: ${resolved}`);
    }
    const raw = fs.readFileSync(resolved, "utf-8");
    const ext = path.extname(filePath).toLowerCase();

    let data: unknown;
    if (ext === ".yaml" || ext === ".yml") {
      // Dynamic import for yaml
      try {
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const yaml = require("js-yaml");
        data = yaml.load(raw);
      } catch {
        throw new Error("js-yaml required for YAML graphs: npm install js-yaml");
      }
    } else {
      data = JSON.parse(raw);
    }

    const parsed = AgentGraphSchema.parse(data);
    return new AgentGraph(parsed);
  }

  /**
   * Create graph from data object.
   */
  static fromData(data: AgentGraphData): AgentGraph {
    return new AgentGraph(AgentGraphSchema.parse(data));
  }
}
